keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,623 @@
+since RBAC allows to one add permissions but not deny them. We won't be able
+to add restrictions on 'non-system' users only for a set of resources using
+RBAC; 'non-system' users here refers to anyone who shouldn't be able to
+update/delete the protected set of CRDs.
+
+Relevant Discussion Thread is sig-apimachinery: https://groups.google.com/g/kubernetes-sig-api-machinery/c/MBa19WTETMQ
+
+#### Story 2: Preventing unnecessarily nested webhooks
+
+Consider a system administrator who would like to apply a set of validations
+to certain requests before handing it off to webhooks defined using frameworks
+like Open Policy Agent.
+
+They would have to nested webhooks within the one added to the auth chain to
+have the intended effect. This enhancement allows the administrator to configure
+this behaviour via a structured API and invoke the additional webhook only when
+relevant. This also allows administrators to define `onError` behaviours for
+separate webhooks, leading to more predictable outcomes.
+
+#### Story 3: Denying requests on certain scenarios
+
+The authorizer chain should be powerful enough to deny anyone making a request
+if certain conditions are satisfied.
+
+#### Story 4: Controlling access of a privileged RBAC role
+
+Certain users associated with a privileged role might need to have their access
+scoped to certain namespaces. Having ordered authorisation modes allows the
+administrator to add a webhook restricting certain user tokens before RBAC
+grants access to the user.
+
+#### Story 5: Varying defaults across versions of the API
+
+If the default values for configuration evolve over time, affected users might
+have to override these values in case they are affected by supplying the flags
+with updated defaults. A configuration format like the one proposed can be
+versioned over time with changing defaults, mitigating this risk by only
+affecting users who knowingly bump the version.
+
+
+### Risks and Mitigations
+
+- In HA clusters, there may be a skew in how the `kube-apiserver` processes in
+each are configured. This may create inconsistencies. Mitigation is to have the
+cluster bootstrapper handle such scenarios.
+- In case an administrator enables this feature and the webhook kubeconfig file
+is invalid or doesn't exist at the specified path, `kube-apiserver` on that node
+will not be able to start. This can be mitigated by fixing the malformed values.
+
+
+## Design Details
+
+We would like to introduce a structured file format which allows authorization
+to be configured using a flag (`--authorization-config-file`) which accepts a
+path to a file on the disk. This feature can be enabled or disabled by the
+explicit feature flag `--authorization-config-from-file`.
+The proposed structure is illustrated below:
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
```

Comment:
Needs filtering and a defined KAS->webhook authn story (that does not continue to abuse the kubeconfig file format).

Reply:
Filtering is added and the authn story is going to be refined later.